backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java

```diff
@@ -15,33 +17,57 @@
 import org.springframework.stereotype.Service;
 
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Collectors;
 
 @Service
 @RequiredArgsConstructor
 public class CustomerServiceImpl implements CustomerService {
 
-    private final CustomerRepository repository;
-    private final CustomerMapper mapper;
+    private final CustomerRepository customerRepository;
+    private final CustomerMapper customerMapper;
 
     @Override
     public Page<CustomerResponse> getCustomersPaginated(final int page, final int size) {
         Pageable pageable = PageRequest.of(page, size);
-        Page<CustomerEntity> customerEntities = repository.findAll(pageable);
+        Page<CustomerEntity> customerEntities = customerRepository.findAll(pageable);
 
         final List<CustomerResponse> customerResponses = customerEntities
                 .getContent()
                 .stream()
-                .map(mapper::convertToDto)
+                .map(customerMapper::convertToDto)
                 .collect(Collectors.toList());
 
         return new PageImpl<>(customerResponses, pageable, customerEntities.getTotalElements());
     }
 
     @Override
     public CustomerResponse getCustomerById(final Long id) {
-        return repository.findById(id)
-                .map(mapper::convertToDto)
+        return customerRepository.findById(id)
+                .map(customerMapper::convertToDto)
                 .orElseThrow(() -> new CustomerNotFoundException(ErrorDict.CUSTOMER_NOT_FOUND));
     }
+
+    @Override
+    public CustomerResponse addCustomer(final CustomerRequest customerRequest) {
+        Optional<CustomerEntity> customerByPhone = customerRepository.findByPhone(customerRequest.getPhone());
+
+        customerByPhone.ifPresent(c -> {
+            if (isDuplicateCustomer(customerRequest, c)) {
+                throw new CustomerException(ErrorDict.CUSTOMER_DUPLICATE);
+            } else {
+                throw new CustomerException(ErrorDict.CUSTOMER_CREATE_PHONE_EXISTS);
+            }
+        });
+
+        CustomerEntity newCustomer = customerMapper.convertToEntity(customerRequest);
+        CustomerEntity savedCustomer = customerRepository.save(newCustomer);
+
+        return customerMapper.convertToDto(savedCustomer);
+    }
+
+    private boolean isDuplicateCustomer(CustomerRequest customerRequest, CustomerEntity c) {
+        return c.getName().equals(customerRequest.getName()) && c.getSurname().equals(customerRequest.getSurname())
+                && c.getAddress().equals(customerRequest.getAddress());
```

Comment:
Troche lipa z tym adresem, może z adresu też zrobić encję?
-country
-city
-postal code
-street name
-street number
-flat number

do tego equals and hashcode

Aktualnie jeśli cały adres jest w Stringu to trzeba pilnować zapisu całego adresu żeby equals przechodził

Reply:
To zadanie już jest w planach, właściwie to ja je sobie przypisałem i chcę to zrobić jak tylko znajdę czas :)